// _@tarojs_components@3.0.14@@tarojs/components/dist/esm/taro-canvas-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './core-de9c9c2d.js';

const LONG_TAP_DELAY = 500;
const Canvas = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.onTouchStart = () => {
            this.timer = setTimeout(() => {
                this.onLongTap.emit();
            }, LONG_TAP_DELAY);
        };
        this.onTouchMove = () => {
            clearTimeout(this.timer);
        };
        this.onTouchEnd = () => {
            clearTimeout(this.timer);
        };
        this.onLongTap = createEvent(this, "longtap", 7);
    }
    render() {
        const { canvasId } = this;
        return (h("canvas", { "canvas-id": canvasId, style: {
                width: '100%',
                height: '100%'
            }, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd }));
    }
    get el() { return getElement(this); }
    static get style() { return "taro-canvas-core{position:relative;display:block;width:300px;height:150px}"; }
};

export { Canvas as taro_canvas_core };
