// _@tarojs_components@3.0.14@@tarojs/components/dist/esm/taro-navigator-core.entry.js
import { r as registerInstance, c as createEvent, h, H as Host } from './core-de9c9c2d.js';
import { c as classNames } from './index-be61bb74.js';

const Taro = require('@tarojs/taro');
const Navigator = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.openType = 'navigate';
        this.isHover = false;
        this.delta = 0;
        this.onSuccess = createEvent(this, "cuccess", 7);
        this.onFail = createEvent(this, "fail", 7);
        this.onComplete = createEvent(this, "Complete", 7);
    }
    onClick() {
        const { openType, onSuccess, onFail, onComplete } = this;
        let promise = Promise.resolve();
        switch (openType) {
            case 'navigate':
                promise = Taro.navigateTo({
                    url: this.url
                });
                break;
            case 'redirect':
                promise = Taro.redirectTo({
                    url: this.url
                });
                break;
            case 'switchTab':
                promise = Taro.switchTab({
                    url: this.url
                });
                break;
            case 'reLaunch':
                promise = Taro.reLaunch({
                    url: this.url
                });
                break;
            case 'navigateBack':
                promise = Taro.navigateBack({
                    delta: this.delta
                });
                break;
            case 'exit':
                promise = Promise.reject(new Error('navigator:fail 暂不支持"openType: exit"'));
                break;
        }
        if (promise) {
            promise.then(res => {
                onSuccess.emit(res);
            }).catch(res => {
                onFail.emit(res);
            }).finally(() => {
                onComplete.emit();
            });
        }
    }
    render() {
        const { isHover, hoverClass } = this;
        return (h(Host, { class: classNames({
                [hoverClass]: isHover
            }) }));
    }
    static get style() { return ".navigator-hover{background:#efefef}"; }
};

export { Navigator as taro_navigator_core };
