// _@tarojs_components@3.0.14@@tarojs/components/dist/esm/taro-progress-core.entry.js
import { r as registerInstance, h, H as Host } from './core-de9c9c2d.js';

const Index = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.percent = 0;
        this.showInfo = false;
        this.borderRadius = 0;
        this.fontSize = 16;
        this.strokeWidth = 6;
        this.activeColor = '#09BB07';
        this.backgroundColor = '#EBEBEB';
        this.active = false;
    }
    render() {
        const { percent, showInfo, borderRadius, fontSize, strokeWidth, activeColor, backgroundColor, active } = this;
        const pgPercent = percent > 100 ? 100 : percent < 0 ? 0 : percent;
        const pgHeight = {
            height: strokeWidth + 'px',
            backgroundColor
        };
        const transition = active ? 'width 1s ease-in-out' : 'none';
        const pgWdith = {
            width: `${pgPercent}%`,
            transition,
            WebkitTransition: transition,
            backgroundColor: activeColor,
            borderRadius: borderRadius ? `${borderRadius}px` : '0px'
        };
        return (h(Host, { class: 'weui-progress' }, h("div", { class: 'weui-progress__bar', style: pgHeight }, h("div", { class: 'weui-progress__inner-bar', style: pgWdith })), showInfo && (h("div", { class: 'weui-progress__opr', style: { 'font-size': `${fontSize}px` } }, h("span", null, pgPercent, "%")))));
    }
};

export { Index as taro_progress_core };
