// _@tarojs_webpack-runner@3.0.14@@tarojs/webpack-runner/dist/config/postcss.conf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPostcssPlugins = void 0;
const autoprefixer = require("autoprefixer");
const path = require("path");
const constparse = require("postcss-plugin-constparse");
const pxtransform = require("postcss-pxtransform");
const resolve_1 = require("resolve");
const helper_1 = require("@tarojs/helper");
const defaultAutoprefixerOption = {
    enable: true,
    config: {
        flexbox: 'no-2009'
    }
};
const defaultPxtransformOption = {
    enable: true,
    config: {
        platform: 'h5'
    }
};
const defaultConstparseOption = {
    constants: [
        {
            key: 'taro-tabbar-height',
            val: '50PX'
        }
    ],
    platform: 'h5'
};
const optionsWithDefaults = ['autoprefixer', 'pxtransform', 'cssModules'];
const plugins = [];
exports.getPostcssPlugins = function (appPath, { designWidth, deviceRatio, postcssOption = {} }) {
    if (designWidth) {
        defaultPxtransformOption.config.designWidth = designWidth;
    }
    if (deviceRatio) {
        defaultPxtransformOption.config.deviceRatio = deviceRatio;
    }
    const autoprefixerOption = helper_1.recursiveMerge({}, defaultAutoprefixerOption, postcssOption.autoprefixer);
    const pxtransformOption = helper_1.recursiveMerge({}, defaultPxtransformOption, postcssOption.pxtransform);
    if (autoprefixerOption.enable) {
        plugins.push(autoprefixer(autoprefixerOption.config));
    }
    if (pxtransformOption.enable) {
        plugins.push(pxtransform(pxtransformOption.config));
    }
    plugins.push(constparse(defaultConstparseOption));
    Object.entries(postcssOption).forEach(([pluginName, pluginOption]) => {
        if (optionsWithDefaults.indexOf(pluginName) > -1)
            return;
        if (!pluginOption || !pluginOption.enable)
            return;
        if (!helper_1.isNpmPkg(pluginName)) {
            // local plugin
            pluginName = path.join(appPath, pluginName);
        }
        try {
            const pluginPath = resolve_1.sync(pluginName, { basedir: appPath });
            plugins.push(require(pluginPath)(pluginOption.config || {}));
        }
        catch (e) {
            const msg = e.code === 'MODULE_NOT_FOUND' ? `缺少postcss插件${pluginName}, 已忽略` : e;
            console.log(msg);
        }
    });
    return plugins;
};
