// _@tarojs_webpack-runner@3.0.14@@tarojs/webpack-runner/dist/config/vue3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customVue3Chain = void 0;
const helper_1 = require("@tarojs/helper");
const chain_1 = require("../util/chain");
function customVue3Chain(chain, config) {
    let vueLoaderPath;
    try {
        vueLoaderPath = require.resolve('vue-loader', {
            paths: [process.cwd()]
        });
    }
    catch (error) {
        console.log(helper_1.chalk.yellow('找不到 vue-loader，请先安装。'));
        process.exit(1);
    }
    const { VueLoaderPlugin } = require(vueLoaderPath);
    const { styleLoaderOption = {} } = config;
    chain.resolve.alias
        .set('vue', '@vue/runtime-dom');
    chain
        .plugin('vueLoaderPlugin')
        .use(VueLoaderPlugin);
    chain.module
        .rule('customStyle')
        .merge({
        use: [{
                loader: 'style-loader',
                options: styleLoaderOption
            }]
    });
    chain.module
        .rule('vue')
        .test(helper_1.REG_VUE)
        .use('vueLoader')
        .loader(vueLoaderPath)
        .options({
        transformAssetUrls: {
            video: ['src', 'poster'],
            'live-player': 'src',
            audio: 'src',
            source: 'src',
            image: 'src',
            'cover-image': 'src',
            'taro-video': ['src', 'poster'],
            'taro-live-player': 'src',
            'taro-audio': 'src',
            'taro-source': 'src',
            'taro-image': 'src',
            'taro-cover-image': 'src'
        },
        compilerOptions: {
            // https://github.com/vuejs/vue-next/blob/master/packages/compiler-core/src/options.ts
            nodeTransforms: [(node) => {
                    if (node.type === 1 /* ELEMENT */) {
                        node = node;
                        const nodeName = node.tag;
                        if (chain_1.DEFAULT_Components.has(nodeName)) {
                            node.tag = `taro-${nodeName}`;
                            node.tagType = 1; /* 0: ELEMENT, 1: COMPONENT */
                        }
                    }
                }]
        }
    });
}
exports.customVue3Chain = customVue3Chain;
