// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-radio-core_2.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-de9c9c2d.js';
var Radio = /** @class */ (function () {
    function Radio(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.value = '';
        this.checked = false;
        this.handleClick = function () {
            if (!_this.checked)
                _this.checked = true;
        };
        this.onChange = createEvent(this, "radiochange", 7);
    }
    Radio.prototype.watchChecked = function (newVal) {
        newVal && this.onChange.emit({ value: this.value });
    };
    Radio.prototype.watchId = function (newVal) {
        if (newVal)
            this.inputEl.setAttribute('id', newVal);
    };
    Radio.prototype.componentDidRender = function () {
        this.id && this.el.removeAttribute('id');
    };
    Radio.prototype.render = function () {
        var _this = this;
        var _a = this, checked = _a.checked, name = _a.name, value = _a.value;
        return (h(Host, { className: 'weui-cells_checkbox', onClick: this.handleClick }, h("input", { ref: function (dom) {
                if (!dom)
                    return;
                _this.inputEl = dom;
                if (_this.id)
                    dom.setAttribute('id', _this.id);
            }, type: 'radio', name: name, value: value, class: 'weui-check', checked: checked, onChange: function (e) { return e.stopPropagation(); } }), h("i", { class: 'weui-icon-checked' }), h("slot", null)));
    };
    Object.defineProperty(Radio.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Radio, "watchers", {
        get: function () {
            return {
                "checked": ["watchChecked"],
                "id": ["watchId"]
            };
        },
        enumerable: true,
        configurable: true
    });
    return Radio;
}());
var RadioGroup = /** @class */ (function () {
    function RadioGroup(hostRef) {
        registerInstance(this, hostRef);
        this.uniqueName = Date.now().toString(36);
        this.onChange = createEvent(this, "change", 7);
    }
    RadioGroup.prototype.function = function (e) {
        e.stopPropagation();
        if (e.target.tagName !== 'TARO-RADIO-CORE')
            return;
        var target = e.target;
        if (target.checked) {
            var childList = this.el.querySelectorAll('taro-radio-core');
            childList.forEach(function (element) {
                if (element !== target) {
                    element.checked = false;
                }
            });
            this.value = e.detail.value;
            this.onChange.emit({
                value: this.value
            });
        }
    };
    RadioGroup.prototype.componentDidLoad = function () {
        var _this = this;
        var childList = this.el.querySelectorAll('taro-radio-core');
        childList.forEach(function (element) {
            element.setAttribute('name', _this.name || _this.uniqueName);
        });
        Object.defineProperty(this.el, 'value', {
            get: function () {
                if (!_this.value) {
                    var childList_1 = _this.el.querySelectorAll('taro-radio-core');
                    _this.value = _this.getValues(childList_1);
                }
                return _this.value;
            },
            configurable: true
        });
    };
    RadioGroup.prototype.getValues = function (childList) {
        var val = '';
        Array.from(childList)
            .forEach(function (element) {
            var checkbox = element.querySelector('input');
            if (checkbox === null || checkbox === void 0 ? void 0 : checkbox.checked) {
                val = checkbox.value || '';
            }
        });
        return val;
    };
    RadioGroup.prototype.render = function () {
        return (h(Host, { class: 'weui-cells_radiogroup' }));
    };
    Object.defineProperty(RadioGroup.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    return RadioGroup;
}());
export { Radio as taro_radio_core, RadioGroup as taro_radio_group_core };
