// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-view-core.entry.js
import { r as registerInstance, c as createEvent, h, H as Host } from './core-de9c9c2d.js';
import { c as classNames } from './index-be61bb74.js';
var View = /** @class */ (function () {
    function View(hostRef) {
        registerInstance(this, hostRef);
        this.hoverStartTime = 50;
        this.hoverStayTime = 400;
        this.hover = false;
        this.touch = false;
        this.startTime = 0;
        this.onLongPress = createEvent(this, "longpress", 7);
    }
    View.prototype.onTouchStart = function () {
        var _this = this;
        if (this.hoverClass) {
            this.touch = true;
            setTimeout(function () {
                if (_this.touch) {
                    _this.hover = true;
                }
            }, this.hoverStartTime);
        }
        this.timeoutEvent = setTimeout(function () {
            _this.onLongPress.emit();
        }, 350);
        this.startTime = Date.now();
    };
    View.prototype.onTouchMove = function () {
        clearTimeout(this.timeoutEvent);
    };
    View.prototype.onTouchEnd = function () {
        var _this = this;
        var spanTime = Date.now() - this.startTime;
        if (spanTime < 350) {
            clearTimeout(this.timeoutEvent);
        }
        if (this.hoverClass) {
            this.touch = false;
            setTimeout(function () {
                if (!_this.touch) {
                    _this.hover = false;
                }
            }, this.hoverStayTime);
        }
    };
    View.prototype.render = function () {
        var _a;
        var cls = classNames((_a = {},
            _a["" + this.hoverClass] = this.hover,
            _a));
        return (h(Host, { class: cls }));
    };
    Object.defineProperty(View, "style", {
        get: function () { return "body,html{-webkit-tap-highlight-color:rgba(0,0,0,0);user-select:none;-moz-user-select:none;-webkit-user-select:none;-ms-user-select:none}taro-view-core{display:block}"; },
        enumerable: true,
        configurable: true
    });
    return View;
}());
export { View as taro_view_core };
