// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-switch-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './core-de9c9c2d.js';
var Switch = /** @class */ (function () {
    function Switch(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.type = 'switch';
        this.checked = false;
        this.color = '#04BE02';
        this.disabled = false;
        this.switchChange = function (e) {
            e.stopPropagation();
            var value = e.target.checked;
            _this.isChecked = value;
            _this.onChange.emit({
                value: value
            });
        };
        this.onChange = createEvent(this, "change", 7);
    }
    Switch.prototype.function = function (newVal, oldVal) {
        if (newVal !== oldVal)
            this.isChecked = newVal;
    };
    Switch.prototype.componentWillLoad = function () {
        this.isChecked = this.checked;
    };
    Switch.prototype.componentDidLoad = function () {
        var _this = this;
        Object.defineProperty(this.el, 'value', {
            get: function () { return _this.isChecked; },
            configurable: true
        });
    };
    Switch.prototype.render = function () {
        var _a = this, type = _a.type, color = _a.color, isChecked = _a.isChecked, name = _a.name, disabled = _a.disabled;
        var style = isChecked
            ? {
                borderColor: color || '04BE02',
                backgroundColor: color || '04BE02'
            }
            : {};
        return (h("input", { type: 'checkbox', class: "weui-" + type, style: style, checked: isChecked, name: name, disabled: disabled, onChange: this.switchChange }));
    };
    Object.defineProperty(Switch.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Switch, "watchers", {
        get: function () {
            return {
                "checked": ["function"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Switch, "style", {
        get: function () { return "taro-switch-core{display:inline-block;width:52px;height:32px}taro-switch-core .weui-switch{display:block;width:100%;height:100%}"; },
        enumerable: true,
        configurable: true
    });
    return Switch;
}());
export { Switch as taro_switch_core };
