// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-checkbox-core_2.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './core-de9c9c2d.js';
var Checkbox = /** @class */ (function () {
    function Checkbox(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.value = '';
        this.checked = false;
        this.handleChange = function (e) {
            e.stopPropagation();
            _this.onChange.emit({
                value: _this.value
            });
        };
        this.onChange = createEvent(this, "checkboxchange", 7);
    }
    Checkbox.prototype.watchId = function (newVal) {
        if (newVal)
            this.inputEl.setAttribute('id', newVal);
    };
    Checkbox.prototype.componentDidRender = function () {
        this.id && this.el.removeAttribute('id');
    };
    Checkbox.prototype.render = function () {
        var _this = this;
        var _a = this, checked = _a.checked, name = _a.name, color = _a.color, value = _a.value;
        return (h(Host, { className: 'weui-cells_checkbox' }, h("input", { ref: function (dom) {
                if (!dom)
                    return;
                _this.inputEl = dom;
                if (_this.id)
                    dom.setAttribute('id', _this.id);
            }, type: 'checkbox', value: value, name: name, class: 'taro-checkbox_checked', style: { color: color }, checked: checked, onChange: this.handleChange }), h("slot", null)));
    };
    Object.defineProperty(Checkbox.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Checkbox, "watchers", {
        get: function () {
            return {
                "id": ["watchId"]
            };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Checkbox, "style", {
        get: function () { return "\@charset \"UTF-8\";.taro-checkbox,.taro-checkbox_checked{position:relative;display:inline-block}.taro-checkbox_checked{min-height:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;outline:0;font-size:23px;border:1px solid #d1d1d1;background-color:#fff;border-radius:3px;color:#1aad19;width:23px;height:23px;vertical-align:0;top:5px}.taro-checkbox_checked:checked:before{font-family:weui;font-style:normal;font-weight:400;font-variant:normal;text-transform:none;text-align:center;speak:none;display:inline-block;vertical-align:middle;text-decoration:inherit;content:\"\";color:inherit;font-size:inherit;position:absolute;top:50%;left:50%;-webkit-transform:translate(-50%,-48%) scale(.73);transform:translate(-50%,-48%) scale(.73)}"; },
        enumerable: true,
        configurable: true
    });
    return Checkbox;
}());
var CheckboxGroup = /** @class */ (function () {
    function CheckboxGroup(hostRef) {
        registerInstance(this, hostRef);
        this.uniqueName = Date.now().toString(36);
        this.onChange = createEvent(this, "change", 7);
    }
    CheckboxGroup.prototype.function = function (e) {
        e.stopPropagation();
        if (e.target.tagName !== 'TARO-CHECKBOX-CORE')
            return;
        var childList = this.el.querySelectorAll('taro-checkbox-core');
        this.value = this.getValues(childList);
        this.onChange.emit({
            value: this.value
        });
    };
    CheckboxGroup.prototype.componentDidLoad = function () {
        var _this = this;
        var childList = this.el.querySelectorAll('taro-checkbox-core');
        childList.forEach(function (element) {
            element.setAttribute('name', _this.name || _this.uniqueName);
        });
        Object.defineProperty(this.el, 'value', {
            get: function () {
                if (!_this.value) {
                    var childList_1 = _this.el.querySelectorAll('taro-checkbox-core');
                    _this.value = _this.getValues(childList_1);
                }
                return _this.value;
            },
            configurable: true
        });
    };
    CheckboxGroup.prototype.getValues = function (childList) {
        return Array.from(childList)
            .filter(function (element) {
            var checkbox = element.querySelector('input');
            return checkbox === null || checkbox === void 0 ? void 0 : checkbox.checked;
        })
            .map(function (element) { return element.value; });
    };
    CheckboxGroup.prototype.render = function () {
        return (h(Host, null));
    };
    Object.defineProperty(CheckboxGroup.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    return CheckboxGroup;
}());
export { Checkbox as taro_checkbox_core, CheckboxGroup as taro_checkbox_group_core };
