// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-input-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './core-de9c9c2d.js';
function getTrueType(type, confirmType, password) {
    if (!type) {
        throw new Error('unexpected type');
    }
    if (confirmType === 'search')
        type = 'search';
    if (password)
        type = 'password';
    if (type === 'digit')
        type = 'number';
    return type;
}
function fixControlledValue(value) {
    return value !== null && value !== void 0 ? value : '';
}
var Input = /** @class */ (function () {
    function Input(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.isOnComposition = false;
        this.onInputExcuted = false;
        this.type = 'text';
        this.password = false;
        this.disabled = false;
        this.maxlength = 140;
        this.autoFocus = false;
        this.confirmType = 'done';
        this.hanldeInput = function (e) {
            e.stopPropagation();
            var _a = _this, type = _a.type, maxlength = _a.maxlength, confirmType = _a.confirmType, password = _a.password;
            if (!_this.isOnComposition && !_this.onInputExcuted) {
                var value = e.target.value;
                var inputType = getTrueType(type, confirmType, password);
                _this.onInputExcuted = true;
                /* 修复 number 类型 maxlength 无效 */
                if (inputType === 'number' && value && maxlength <= value.length) {
                    value = value.substring(0, maxlength);
                    e.target.value = value;
                }
                // 修复 IOS 光标跳转问题
                // if (!(['number', 'file'].indexOf(inputType) >= 0)) {
                //   const pos = e.target.selectionEnd
                //   setTimeout(
                //     () => {
                //       e.target.selectionStart = pos
                //       e.target.selectionEnd = pos
                //     }
                //   )
                // }
                _this.onInput.emit({
                    value: value,
                    cursor: value.length
                });
            }
        };
        this.handleFocus = function (e) {
            _this.onInputExcuted = false;
            _this.onFocus.emit({
                value: e.target.value
            });
        };
        this.handleBlur = function (e) {
            _this.onBlur.emit({
                value: e.target.value
            });
        };
        this.handleChange = function (e) {
            e.stopPropagation();
            _this.onChange.emit({
                value: e.target.value
            });
        };
        this.handleKeyDown = function (e) {
            var value = e.target.value;
            _this.onInputExcuted = false;
            e.stopPropagation();
            _this.onKeyDown.emit({ value: value });
            e.keyCode === 13 && _this.onConfirm.emit({ value: value });
        };
        this.handleComposition = function (e) {
            if (!(e.target instanceof HTMLInputElement))
                return;
            if (e.type === 'compositionend') {
                _this.isOnComposition = false;
                _this.onInput.emit({ value: e.target.value });
            }
            else {
                _this.isOnComposition = true;
            }
        };
        this.onInput = createEvent(this, "input", 7);
        this.onFocus = createEvent(this, "focus", 7);
        this.onBlur = createEvent(this, "blur", 7);
        this.onConfirm = createEvent(this, "confirm", 7);
        this.onChange = createEvent(this, "change", 7);
        this.onKeyDown = createEvent(this, "keydown", 7);
    }
    Input.prototype.componentDidLoad = function () {
        var _this = this;
        if (this.type === 'file') {
            this.fileListener = function () {
                _this.onInput.emit();
            };
            this.inputRef.addEventListener('change', this.fileListener);
        }
        else {
            this.inputRef.addEventListener('compositionstart', this.handleComposition);
            this.inputRef.addEventListener('compositionend', this.handleComposition);
        }
        Object.defineProperty(this.el, 'value', {
            get: function () { return _this.inputRef.value; },
            set: function (value) { return (_this.value = value); },
            configurable: true
        });
    };
    Input.prototype.componentDidUnload = function () {
        if (this.type === 'file') {
            this.inputRef.removeEventListener('change', this.fileListener);
        }
    };
    Input.prototype.render = function () {
        var _this = this;
        var _a = this, value = _a.value, type = _a.type, password = _a.password, placeholder = _a.placeholder, disabled = _a.disabled, maxlength = _a.maxlength, autoFocus = _a.autoFocus, confirmType = _a.confirmType, name = _a.name;
        return (h("input", { ref: function (input) {
                _this.inputRef = input;
                autoFocus && (input === null || input === void 0 ? void 0 : input.focus());
            }, class: 'weui-input', value: fixControlledValue(value), type: getTrueType(type, confirmType, password), placeholder: placeholder, disabled: disabled, maxlength: maxlength, autofocus: autoFocus, name: name, onInput: this.hanldeInput, onFocus: this.handleFocus, onBlur: this.handleBlur, onChange: this.handleChange, onKeyDown: this.handleKeyDown }));
    };
    Object.defineProperty(Input.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Input, "style", {
        get: function () { return "input,taro-input-core{display:block}input{height:1.4rem;text-align:inherit;text-overflow:clip;overflow:hidden;white-space:nowrap}"; },
        enumerable: true,
        configurable: true
    });
    return Input;
}());
export { Input as taro_input_core };
