// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-canvas-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './core-de9c9c2d.js';
var LONG_TAP_DELAY = 500;
var Canvas = /** @class */ (function () {
    function Canvas(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.onTouchStart = function () {
            _this.timer = setTimeout(function () {
                _this.onLongTap.emit();
            }, LONG_TAP_DELAY);
        };
        this.onTouchMove = function () {
            clearTimeout(_this.timer);
        };
        this.onTouchEnd = function () {
            clearTimeout(_this.timer);
        };
        this.onLongTap = createEvent(this, "longtap", 7);
    }
    Canvas.prototype.render = function () {
        var canvasId = this.canvasId;
        return (h("canvas", { "canvas-id": canvasId, style: {
                width: '100%',
                height: '100%'
            }, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd }));
    };
    Object.defineProperty(Canvas.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Canvas, "style", {
        get: function () { return "taro-canvas-core{position:relative;display:block;width:300px;height:150px}"; },
        enumerable: true,
        configurable: true
    });
    return Canvas;
}());
export { Canvas as taro_canvas_core };
