// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-audio-core.entry.js
import { r as registerInstance, c as createEvent, h } from './core-de9c9c2d.js';
var Audio = /** @class */ (function () {
    function Audio(hostRef) {
        registerInstance(this, hostRef);
        this.controls = true;
        this.autoplay = false;
        this.loop = false;
        this.muted = false;
        this.onError = createEvent(this, "error", 7);
        this.onPlay = createEvent(this, "play", 7);
        this.onPause = createEvent(this, "pause", 7);
        this.onTimeUpdate = createEvent(this, "timeupdate", 7);
        this.onEnded = createEvent(this, "ended", 7);
    }
    Audio.prototype.bindevent = function () {
        var _this = this;
        this.audio.addEventListener('timeupdate', function (e) {
            _this.onTimeUpdate.emit({
                duration: e.srcElement.duration,
                currentTime: e.srcElement.duration
            });
        });
        this.audio.addEventListener('ended', function () {
            _this.onEnded.emit();
        });
        this.audio.addEventListener('play', function () {
            _this.onPlay.emit();
        });
        this.audio.addEventListener('pause', function (e) {
            _this.onPause.emit(e);
        });
        // 1网络错误, 2解码错误, 3解码错误，4 不合适资源
        this.audio.addEventListener('error', function (e) {
            var _a;
            _this.onError.emit({
                errMsg: (_a = e.srcElement.error) === null || _a === void 0 ? void 0 : _a.code
            });
        });
    };
    Audio.prototype.componentDidLoad = function () {
        this.bindevent();
    };
    Audio.prototype.render = function () {
        var _this = this;
        var _b = this, src = _b.src, controls = _b.controls, autoplay = _b.autoplay, loop = _b.loop, muted = _b.muted;
        return (h("audio", { src: src, controls: controls, autoplay: autoplay, loop: loop, muted: muted, ref: function (audio) { _this.audio = audio; } }));
    };
    Object.defineProperty(Audio, "style", {
        get: function () { return "audio{max-width:100%;outline:none}"; },
        enumerable: true,
        configurable: true
    });
    return Audio;
}());
export { Audio as taro_audio_core };
