// _@tarojs_components@3.0.14@@tarojs/components/dist/esm-es5/taro-progress-core.entry.js
import { r as registerInstance, h, H as Host } from './core-de9c9c2d.js';
var Index = /** @class */ (function () {
    function Index(hostRef) {
        registerInstance(this, hostRef);
        this.percent = 0;
        this.showInfo = false;
        this.borderRadius = 0;
        this.fontSize = 16;
        this.strokeWidth = 6;
        this.activeColor = '#09BB07';
        this.backgroundColor = '#EBEBEB';
        this.active = false;
    }
    Index.prototype.render = function () {
        var _a = this, percent = _a.percent, showInfo = _a.showInfo, borderRadius = _a.borderRadius, fontSize = _a.fontSize, strokeWidth = _a.strokeWidth, activeColor = _a.activeColor, backgroundColor = _a.backgroundColor, active = _a.active;
        var pgPercent = percent > 100 ? 100 : percent < 0 ? 0 : percent;
        var pgHeight = {
            height: strokeWidth + 'px',
            backgroundColor: backgroundColor
        };
        var transition = active ? 'width 1s ease-in-out' : 'none';
        var pgWdith = {
            width: pgPercent + "%",
            transition: transition,
            WebkitTransition: transition,
            backgroundColor: activeColor,
            borderRadius: borderRadius ? borderRadius + "px" : '0px'
        };
        return (h(Host, { class: 'weui-progress' }, h("div", { class: 'weui-progress__bar', style: pgHeight }, h("div", { class: 'weui-progress__inner-bar', style: pgWdith })), showInfo && (h("div", { class: 'weui-progress__opr', style: { 'font-size': fontSize + "px" } }, h("span", null, pgPercent, "%")))));
    };
    return Index;
}());
export { Index as taro_progress_core };
