// _@tarojs_components@3.0.14@@tarojs/components/dist/cjs/taro-view-core.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-06196d96.js');
const index$1 = require('./index-00392f71.js');

const View = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        this.hoverStartTime = 50;
        this.hoverStayTime = 400;
        this.hover = false;
        this.touch = false;
        this.startTime = 0;
        this.onLongPress = core.createEvent(this, "longpress", 7);
    }
    onTouchStart() {
        if (this.hoverClass) {
            this.touch = true;
            setTimeout(() => {
                if (this.touch) {
                    this.hover = true;
                }
            }, this.hoverStartTime);
        }
        this.timeoutEvent = setTimeout(() => {
            this.onLongPress.emit();
        }, 350);
        this.startTime = Date.now();
    }
    onTouchMove() {
        clearTimeout(this.timeoutEvent);
    }
    onTouchEnd() {
        const spanTime = Date.now() - this.startTime;
        if (spanTime < 350) {
            clearTimeout(this.timeoutEvent);
        }
        if (this.hoverClass) {
            this.touch = false;
            setTimeout(() => {
                if (!this.touch) {
                    this.hover = false;
                }
            }, this.hoverStayTime);
        }
    }
    render() {
        const cls = index$1.classNames({
            [`${this.hoverClass}`]: this.hover
        });
        return (core.h(core.Host, { class: cls }));
    }
    static get style() { return "body,html{-webkit-tap-highlight-color:rgba(0,0,0,0);user-select:none;-moz-user-select:none;-webkit-user-select:none;-ms-user-select:none}taro-view-core{display:block}"; }
};

exports.taro_view_core = View;
