// _@tarojs_components@3.0.14@@tarojs/components/dist/cjs/taro-canvas-core.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-06196d96.js');

const LONG_TAP_DELAY = 500;
const Canvas = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        this.onTouchStart = () => {
            this.timer = setTimeout(() => {
                this.onLongTap.emit();
            }, LONG_TAP_DELAY);
        };
        this.onTouchMove = () => {
            clearTimeout(this.timer);
        };
        this.onTouchEnd = () => {
            clearTimeout(this.timer);
        };
        this.onLongTap = core.createEvent(this, "longtap", 7);
    }
    render() {
        const { canvasId } = this;
        return (core.h("canvas", { "canvas-id": canvasId, style: {
                width: '100%',
                height: '100%'
            }, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd }));
    }
    get el() { return core.getElement(this); }
    static get style() { return "taro-canvas-core{position:relative;display:block;width:300px;height:150px}"; }
};

exports.taro_canvas_core = Canvas;
