// _@tarojs_components@3.0.14@@tarojs/components/dist/cjs/taro-switch-core.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-06196d96.js');

const Switch = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        this.type = 'switch';
        this.checked = false;
        this.color = '#04BE02';
        this.disabled = false;
        this.switchChange = e => {
            e.stopPropagation();
            const value = e.target.checked;
            this.isChecked = value;
            this.onChange.emit({
                value
            });
        };
        this.onChange = core.createEvent(this, "change", 7);
    }
    function(newVal, oldVal) {
        if (newVal !== oldVal)
            this.isChecked = newVal;
    }
    componentWillLoad() {
        this.isChecked = this.checked;
    }
    componentDidLoad() {
        Object.defineProperty(this.el, 'value', {
            get: () => this.isChecked,
            configurable: true
        });
    }
    render() {
        const { type, color, isChecked, name, disabled } = this;
        const style = isChecked
            ? {
                borderColor: color || '04BE02',
                backgroundColor: color || '04BE02'
            }
            : {};
        return (core.h("input", { type: 'checkbox', class: `weui-${type}`, style: style, checked: isChecked, name: name, disabled: disabled, onChange: this.switchChange }));
    }
    get el() { return core.getElement(this); }
    static get watchers() { return {
        "checked": ["function"]
    }; }
    static get style() { return "taro-switch-core{display:inline-block;width:52px;height:32px}taro-switch-core .weui-switch{display:block;width:100%;height:100%}"; }
};

exports.taro_switch_core = Switch;
