// _@tarojs_components@3.0.14@@tarojs/components/dist/cjs/taro-progress-core.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-06196d96.js');

const Index = class {
    constructor(hostRef) {
        core.registerInstance(this, hostRef);
        this.percent = 0;
        this.showInfo = false;
        this.borderRadius = 0;
        this.fontSize = 16;
        this.strokeWidth = 6;
        this.activeColor = '#09BB07';
        this.backgroundColor = '#EBEBEB';
        this.active = false;
    }
    render() {
        const { percent, showInfo, borderRadius, fontSize, strokeWidth, activeColor, backgroundColor, active } = this;
        const pgPercent = percent > 100 ? 100 : percent < 0 ? 0 : percent;
        const pgHeight = {
            height: strokeWidth + 'px',
            backgroundColor
        };
        const transition = active ? 'width 1s ease-in-out' : 'none';
        const pgWdith = {
            width: `${pgPercent}%`,
            transition,
            WebkitTransition: transition,
            backgroundColor: activeColor,
            borderRadius: borderRadius ? `${borderRadius}px` : '0px'
        };
        return (core.h(core.Host, { class: 'weui-progress' }, core.h("div", { class: 'weui-progress__bar', style: pgHeight }, core.h("div", { class: 'weui-progress__inner-bar', style: pgWdith })), showInfo && (core.h("div", { class: 'weui-progress__opr', style: { 'font-size': `${fontSize}px` } }, core.h("span", null, pgPercent, "%")))));
    }
};

exports.taro_progress_core = Index;
